packages/material-ui-lab/src/DataGrid/DataGrid.js

```diff
@@ -346,15 +359,84 @@ const DataGrid = React.forwardRef(function DataGrid(props, ref) {
     }
   };
 
+  const [isLoading, setLoading] = React.useState(loading);
+
+  React.useEffect(() => {
+    setLoading(loading);
+  }), [loading];
+
   const [data, setData] = React.useState([]);
 
+  const [paginationState, setPaginationState] = React.useState({
+    pagination,
+    paginationPage,
+    paginationPageSize,
+    paginationRowsPerPageOptions,
+    paginationKey
+  });
+
   React.useEffect(() => {
-    const newData = dataProvider.getList({
-      sorting,
+    setPaginationState({
+      pagination,
+      paginationPage,
+      paginationPageSize,
+      paginationRowsPerPageOptions,
+      paginationKey
     });
 
-    setData(newData);
-  }, [dataProvider, sorting]);
+  }, [pagination, paginationPage, paginationPageSize, paginationRowsPerPageOptions, paginationKey]);
+
+  const loadAdditionalData = async () => {
+    const nextPaginationKey = await dataProvider.loadMoreRows(pagination.paginationKey);
+
+    setPaginationState(prevPagination => ({ ...prevPagination, paginationKey: nextPaginationKey }));
+  }
+
+  React.useEffect(() => { // Loads new Data, if the last page is reached and it is not already loading
+    if ((paginationState.paginationPage + 1) * paginationState.paginationPageSize >= rowsData.length && dataProvider.loadMoreRows && !isLoading) {
+      loadAdditionalData();
+    }
+  }, [paginationState, rowsData, dataProvider, isLoading]);
+
+  const handlePageChange = (event, page) => {
+
+    setPaginationState(prevPagination => ({ ...prevPagination, paginationPage: page }));
+
+    if (onPageChange) {
+      onPageChange(event, page);
+    }
+  };
+
+  const handlePagsizeChange = event => {
+    const pageSize = event.target.value;
+    let page = paginationState.paginationPage;
+    if (rowsData.length < pageSize * page) {
+      // Tries to show the same rows as before with then new page size
+      page = rowsData.length === pageSize ? 0 : Math.ceil(rowsData.length / pageSize) - 1;
+      handlePageChange(event, page);
+    }
+    setPaginationState(prevPagination => ({ ...prevPagination, paginationPageSize: pageSize, paginationPage: page }));
+    if (onRowsPerPageChange) {
+      onRowsPerPageChange(event, pageSize);
+    }
+  };
+
+  React.useEffect(() => {
+
+    const getNewData = async () => {
+
+      const loadingTimer = setTimeout(() => setLoading(true), 500);
+
+      const newData = await dataProvider.getList({
+        sorting,
+        pagination: paginationState
+      });
+      setData(newData);
+      clearTimeout(loadingTimer);
+      setLoading(loading)
+    };
```

Comment:
need to clear the timeout in the effect cleanup method.

Reply:
Of course